File: Algorithms/p108_Convert_Sorted_Array_to_Binary_Search_Tree/p108_Convert_Sorted_Array_to_Binary_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return sortedArrayToBSTStep(nums,0,nums.size()-1);
    }
    TreeNode* sortedArrayToBSTStep(vector<int>& nums, int i, int j){
        if(i<=j){
            int mid = (i+j+1)/2;
            TreeNode* root = new TreeNode(nums[mid]);
            root->left = sortedArrayToBSTStep(nums,i,mid-1);
            root->right = sortedArrayToBSTStep(nums,mid+1,j);
            return root;
        }else{
            return 0;
        }
    }
};

File: Algorithms/p108_Convert_Sorted_Array_to_Binary_Search_Tree/p108_Convert_Sorted_Array_to_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        return self.sortedArrayToBSTStep(nums,0,len(nums)-1)
        
    def sortedArrayToBSTStep(self, nums, i, j):
        if i<=j:
            mid = (i+j)//2
            root = TreeNode(nums[mid])
            root.left = self.sortedArrayToBSTStep(nums,i,mid-1);
            root.right = self.sortedArrayToBSTStep(nums,mid+1,j);
            return root
        else:
            return None
